### ISSUE TITLE
Learning rate doubly defined for loaded model

### ISSUE THREAD

**Issue by JRMeyer** [mozilla/TTS#256]
Hi @erogol et al.

In `train.py`, the Adam optimizer and learning rates for all parameters are set as such [[1](https://github.com/mozilla/TTS/blob/b1657d70b160c097be7ac24290514a9aed010452/train.py#L467)]:

```
optimizer = optim.Adam(model.parameters(), lr=c.lr, weight_decay=0)
```

However, if you reload a model, then the learning rates seem to be reset for the optimizer as such [[2]](https://github.com/mozilla/TTS/blob/b1657d70b160c097be7ac24290514a9aed010452/train.py#L495):

```
for group in optimizer.param_groups:
    group['lr'] = c.lr
```

Is there a reason we have to set the learning rate again after loading in a pre-trained model?

Thanks
-josh

**Comment by JRMeyer** [comment]
@twerkmeister -- any thoughts on this?

**Comment by twerkmeister** [comment]
Good question! My hunch is, that it is because some parts of the model don't get restored but recreated from scratch when you restart a training with changed parameters. An example for this is the prenet and final projection component in Tacotron2 when you restart with a different `r` value. The number incoming/outgoing connections changes and so that network part gets regenerated. Possibly it resets the learning rate then to make sure the whole network is being readjusted to this new component.

**Comment by erogol** [comment]
@JRMeyer the reason was that if you like to finetune the model than you might prefer smaller LR than before. So it is important to reset it that way.

**Comment by JRMeyer** [comment]
@erogol -- does this mean that when you load an old model, any `optimizer` params which were defined earlier in `main()` get re-written?

My goal is to experiment with different learning rates for different layers during fine-tuning. To be exact, below is the relevant section of my `main()` function from `train.py`.

Are the learning rates defined in `optimizer` the ones used in training? Monitoring `tensorboard`'s stats, it seems my code works as intended, but your answer makes me uncertain now.

```
def main(args):
    # DISTRUBUTED                                                                                                                                                                                                                                                                                                                                                                                                                        
    if num_gpus > 1:
        init_distributed(args.rank, num_gpus, args.group_id,
                         c.distributed["backend"], c.distributed["url"])
    num_chars = len(phonemes) if c.use_phonemes else len(symbols)
    model = setup_model(num_chars, c)

    print(" | > Num output units : {}".format(ap.num_freq), flush=True)

    # optimizer = optim.Adam(                                                                                                                                                                                                                                                                                                                                                                                                            
    #     model.parameters(),                                                                                                                                                                                                                                                                                                                                                                                                            
    #     lr=c.lr,                                                                                                                                                                                                                                                                                                                                                                                                                       
    #     weight_decay=0                                                                                                                                                                                                                                                                                                                                                                                                                 
    # )                                                                                                                                                                                                                                                                                                                                                                                                                                  
    optimizer = optim.Adam(
        [
            {"params": model.embedding.parameters(), "lr": c.lr_embedding},
            {"params": model.encoder.parameters(), "lr": c.lr_encoder},
            {"params": [param for name, param in model.decoder.named_parameters() if "prenet" in name], "lr": c.lr_prenet},
            {"params": [param for name, param in model.decoder.named_parameters() if "attention" in name], "lr": c.lr_attention},
            {"params": [param for name, param in model.decoder.named_parameters() if "decoder_rnn" in name], "lr": c.lr_decoder_rnn},
            {"params": [param for name, param in model.decoder.named_parameters() if "linear_projection" in name], "lr": c.lr_decoder_projection},
            {"params": [param for name, param in model.decoder.named_parameters() if "stopnet" in name], "lr": c.lr_stopnet},
            {"params": [param for name, param in model.decoder.named_parameters() if "go_frame" in name], "lr": c.lr_go_frame},
            {"params": model.postnet.parameters(), "lr": c.lr_postnet}
        ],
        weight_decay=0
    )
    optimizer_st = optim.Adam(
        model.decoder.stopnet.parameters(), lr=c.lr, weight_decay=0)

    if c.loss_masking:
        criterion = L1LossMasked() if c.model == "Tacotron" else MSELossMasked()
    else:
        criterion = nn.L1Loss() if c.model == "Tacotron" else nn.MSELoss()
    criterion_st = nn.BCEWithLogitsLoss()

    if args.restore_path:
        checkpoint = torch.load(args.restore_path)
        try:
            # TODO: fix optimizer init, model.cuda() needs to be called before                                                                                                                                                                                                                                                                                                                                                           
            # optimizer restore                                                                                                                                                                                                                                                                                                                                                                                                          
            # optimizer.load_state_dict(checkpoint['optimizer'])                                                                                                                                                                                                                                                                                                                                                                         
            if len(c.reinit_layers) > 0:
                raise RuntimeError
            model.load_state_dict(checkpoint['model'])
        except:
            print(" > Partial model initialization.")
            partial_init_flag = True
            model_dict = model.state_dict()
            model_dict = set_init_dict(model_dict, checkpoint, c)
            model.load_state_dict(model_dict)
            del model_dict
        if use_cuda:
            model = model.cuda()
            criterion.cuda()
            criterion_st.cuda()
        # for group in optimizer.param_groups:                                                                                                                                                                                                                                                                                                                                                                                           
        #     group['lr'] = c.lr                                                                                                                                                                                                                                                                                                                                                                                                         
        print(
            " > Model restored from step %d" % checkpoint['step'], flush=True)
        start_epoch = checkpoint['epoch']
        # best_loss = checkpoint['postnet_loss']                                                                                                                                                                                                                                                                                                                                                                                         
        args.restore_step = checkpoint['step']
    else:
        args.restore_step = 0
        if use_cuda:
            model = model.cuda()
            criterion.cuda()
            criterion_st.cuda()
```

**Comment by erogol** [comment]
That should work just fine since you don't load the checkpoint's optimizer
